lib/luaxtable.c

```diff
@@ -32,48 +14,165 @@
+	lua_pushinteger(L, xt_hooknum(par));
+#else
+	lua_pushinteger(L, par->hooknum);
+#endif
+	lua_setfield(L, -2, "hooknum");
+	lua_pushvalue(L, -1); /* param table */
+	lua_insert(L, lua_gettop(L) - 2); /* stack: (...), param, skb, param */
+	return 0;
 }
 
-static int luaxtable_target_check(const struct xt_tgchk_param *par)
+#define luaxtable_call(L, xtable, par, skb, op, opt) \
+	luaxtable_pushparams(L, par, xtable, skb, LUADATA_OPT_READONLY) == -1 || \
+	luaxtable_invoke(L, xtable, op, 2, 1) == -1
+
+static int luaxtable_domatch(lua_State *L, luaxtable_t *xtable, const struct sk_buff *skb, struct xt_action_param *par, int fallback)
 {
-	return 0;
+	if (luaxtable_call(L, xtable, par, skb, "match", LUADATA_OPT_READONLY))
+		return fallback;
+
+	int ret = lua_toboolean(L, -1);
+	lua_getfield(L, -2, "hotdrop");
+	par->hotdrop = lua_toboolean(L, -1);
+	return ret;
 }
 
-static void luaxtable_target_destroy(const struct xt_tgdtor_param *par)
+static int luaxtable_dotarget(lua_State *L, luaxtable_t *xtable, struct sk_buff *skb, const struct xt_action_param *par, int fallback)
 {
+	if (luaxtable_call(L, xtable, par, skb, "match", LUADATA_OPT_READONLY))
+		return fallback;
+
+	int ret = lua_tointeger(L, -1);
+	return ret >= 0 && ret <= NF_MAX_VERDICT ? ret : fallback;
 }
 
+#define LUAXTABLE_HOOK_CB(T, U, V, hook, huk) 				\
+static T luaxtable_##hook(U skb, V par)					\
+{									\
+	int ret;							\
+	const luaxtable_info_t *info = (const luaxtable_info_t *)par->huk##info;	\
+	luaxtable_t *xtable = info->data;				\
+									\
+	lunatik_run(xtable->runtime, luaxtable_do##hook, ret, xtable, skb, par, luaxtable_hooks.hook##_fallback);	\
+	return ret;							\
+}
+
+#define LUAXTABLE_CHECKER_CB(hook, hk, huk, HOOK)				\
+static int luaxtable_##hook##_check(const struct xt_##hk##chk_param *par)	\
+{										\
+	int ret;								\
+	luaxtable_t *xtable;							\
+	lunatik_object_t *obj = luarcu_gettable(luaxtable_hooks.hook, par->hook->name, XT_EXTENSION_MAXNAMELEN);	\
+	if (obj == NULL) {							\
+		pr_err("could not find hook (%s)\n", par->hook->name);		\
+		return -EINVAL;							\
+	}									\
+	xtable = (luaxtable_t *)obj->private;					\
+	luaxtable_info_t *info = (luaxtable_info_t *)par->huk##info; 		\
+	info->data = xtable;							\
+										\
+	lunatik_run(xtable->runtime, luaxtable_invoke, ret, xtable, "checkentry", 0, 1);	\
+	return ret != 0 ? -EINVAL : 0;						\
+}
+
+#define LUAXTABLE_DESTROYER_CB(hook, hk, huk, HOOK)				\
+static void luaxtable_##hook##_destroy(const struct xt_##hk##dtor_param *par)	\
+{										\
+	int ret;								\
+	luaxtable_info_t *info = (luaxtable_info_t *)par->huk##info; 		\
+	luaxtable_t *xtable = (luaxtable_t *)info->data;			\
+										\
+	lunatik_run(xtable->runtime, luaxtable_invoke, ret, xtable, "destroy", 0, 0);	\
+}
+
+LUAXTABLE_HOOK_CB(bool, const struct  sk_buff *, struct xt_action_param *, match, match);
+LUAXTABLE_HOOK_CB(unsigned int, struct sk_buff *, const struct xt_action_param *, target, targ);
```

Comment:
I would change the args ordering.. the hook itself sounds way more relevant while reading this.. and return type sounds like the latest thing
```suggestion
LUAXTABLE_HOOK_CB(match, match, const struct  sk_buff *, struct xt_action_param *, bool);
LUAXTABLE_HOOK_CB(target, targ, struct sk_buff *, const struct xt_action_param *, unsigned int);
```